### PR TITLE
fix unit-consistency

### PR DESCRIPTION
**Pull request by naomi-simumba**
addresses issue https://github.com/gridfm/gridfm-graphkit/issues/98 as suggested in the issue

### CODE REVIEW

**Comment by romeokienzler** [comment]
@naomi-simumba 

1. did u run the pre-commit? @albanpuech  are we now running the pre-commit on every PR?
2. mind adding a regression test for the angle-violation computation?
3. should we add this change to the change log? - I think we need to bump the minor version (the y part in vX.Y.Z) so previous runs might not be compatible

**Comment by albanpuech** [comment]
Thank you @naomi-simumba and @romeokienzler 


> @naomi-simumba
> 
> 1. did u run the pre-commit? @albanpuech  are we now running the pre-commit on every PR?
> 2. mind adding a regression test for the angle-violation computation?
> 3. should we add this change to the change log? - I think we need to bump the minor version (the y part in vX.Y.Z) so previous runs might not be compatible

@romeokienzler:
1. Yes, we are now running it on every PR
3. What do you mean "so previous runs might not be compatible" ?

**Comment by romeokienzler** [comment]
> What do you mean "so previous runs might not be compatible" ?
@albanpuech 

I meant reported metric values might not be compatible

_It does change the value of one reported test metric (previously inflated ~57×). A branch_angle_violation_mean from before #99 is not comparable to one after. → Numerical comparability of that one metric breaks._

**Comment by albanpuech** [comment]
> > What do you mean "so previous runs might not be compatible" ?
> > @albanpuech
> 
> I meant reported metric values might not be compatible
> 
> _It does change the value of one reported test metric (previously inflated ~57×). A branch_angle_violation_mean from before #99 is not comparable to one after. → Numerical comparability of that one metric breaks._

Let's run the integration test and double-check, but it should be compatible. The angle difference violations were 0 before, and I think it will remain 0 even with the fix. At least that's what I got so far with all the models I have trained. Let's see if that is also the case with so few training samples and epochs :)

**Comment by romeokienzler** [comment]
Hi @naomi-simumba, gentle ping on this one — just checking in to see whether you've had a chance to look at the earlier feedback. No rush, but let me know if anything's blocking you. Thanks!

**Comment by naomi-simumba** [comment]
Thank you, both!

To make it testable, I extracted the angle violation block into a standalone function` compute_angle_violation(`) in `tasks/utils.py`. `test_step()` now calls that function directly, so behaviour is unchanged. The benefit is that the function can be imported and called in tests with controlled synthetic inputs. With that in place, I added `tests/test_violations.py` with two regression tests:
- one confirming no false positives when angle differences are within limits
- one that will break if the `* torch.pi / 180.0` unit conversion is ever reverted, which is the actual regression guard for this fix

Also, @romeokienzler , I have now done the pre-commit.

**Comment by romeokienzler** [comment]
@naomi-simumba thanks for this — the refactor into `compute_angle_violation()` with a clear Google-style docstring plus the two regression tests in `tests/test_violations.py` (especially the one guarding the degrees→radians conversion) directly answers the earlier review points. This looks basically ready.

**What's needed**
- **Rebase onto latest `main`** — GitHub reports the branch as `BEHIND`, and CONTRIBUTING asks for a rebase before merge.
- **`pip-audit` check** — this is failing, but it's the known repo-wide infra blocker (PYSEC-2026-3624 / lightning) being fixed separately, **not** anything in your PR. All your other checks (DCO, pytests, pre-commit) are green.
- **Version bump (maintainer call):** the thread raised whether to bump the minor version given the metric-value change. Happy to defer to @romeokienzler / @albanpuech on whether that's needed here.

Nice, focused change — a maintainer will follow up.

 — 🤖 _automated pre-review; a maintainer will follow up_

File: gridfm_graphkit/tasks/utils.py [hunk 1]
```diff
@@ -1,11 +1,58 @@
...
+
+    from_bus = bus_edge_index[0]
+    to_bus = bus_edge_index[1]
+    angle_diff = bus_angles[from_bus] - bus_angles[to_bus]
+    angle_diff = (angle_diff + torch.pi) % (2 * torch.pi) - torch.pi  # wrap to [-π, π]
+
+    angle_excess_low = F.relu(angle_min - angle_diff)
+    angle_excess_high = F.relu(angle_diff - angle_max)
+
+    return torch.mean(angle_excess_low + angle_excess_high)
```
Comment (romeokienzler) [inline gridfm_graphkit/tasks/utils.py:58]: 🟡 **Denominator inconsistency (pre-existing, re-exposed here).** This averages over *all* edges — both branch directions, i.e. 2×num_branches — whereas thermal violations in the same `test_step` split forward/reverse at `half_edges`. If forward/reverse angle limits ever differ, or downstream code assumes the angle mean is per-branch-normalized like the thermal means, the two metrics use different denominators. Not introduced by this PR (moved code), but worth a comment or aligning the denominator while it's being touched.

**Review by romeokienzler — COMMENTED** [review]
## Review: core fix is correct ✅

Converting the degree-valued `ANG_MIN`/`ANG_MAX` limits to radians so they compare against the radian bus angles is the right fix for #98. Verified against `normalizers.py`: after `inverse_transform`, `VA` stays in **radians** while `ANG_MIN`/`ANG_MAX` are restored to **degrees**, and `VA_OUT` isn't rescaled — so the direction is sound. Extracting `compute_angle_violation` + adding `test_violations.py` is a nice touch.

Two findings on a file this PR doesn't touch, plus one inline nit. The first is worth folding into this PR.

### 🔴 1. Same unit bug left unfixed in the AC/DC baseline — `gridfm_graphkit/tasks/opf_ac_dc_baseline.py:205`
`opf_ac_dc_baseline.py` has the *exact* radian-vs-degree inconsistency this PR fixes, but isn't updated here:
- `Va` is converted degrees→radians (lines 189-190), but `br['ang_min']`/`br['ang_max']` come straight from `branch_data.parquet` (MATPOWER `ANGMIN`/`ANGMAX`, in **degrees**).
- At lines 205-206 / 211-212, `np.maximum(ang_min - ac_angle_diff, 0)` subtracts a radian diff (~0.1) from a degree limit (~−30), so **real violations are essentially never detected** and the reported *"AC/DC Mean branch angle difference violation (radians)"* baseline is wrong.
- Net effect: this PR fixes the model path but leaves the **reference baseline** buggy, so model-vs-baseline angle-violation comparisons become apples-to-oranges — which partly defeats the purpose. Recommend fixing it in this same PR.

### 🟡 2. Duplicated logic — `gridfm_graphkit/tasks/opf_ac_dc_baseline.py:201`
The baseline reimplements the same wrap-to-[−π,π] + ReLU-excess logic in numpy instead of reusing the new `compute_angle_violation` helper. This duplication is *why* the baseline instance (finding 1) was missed. Consolidating the unit-conversion convention into one place would prevent future divergence.

### Housekeeping
Branch is behind `main`; a rebase clears the stale `pip-audit` failure (fixed by #107) and the `pytests` breakage (fixed by #109).

**Comment by romeokienzler** [comment]
@naomi-simumba nice — this new commit (`5120b7f`) carries the same degrees→radians fix into the AC/DC baseline path (`_compute_branch_violations` in `opf_ac_dc_baseline.py`), which the earlier `opf_task.py` refactor didn't touch. Extracting `ang_min_rad`/`ang_max_rad` and the two matching baseline regression tests in `tests/test_violations.py` (the AC + DC degrees-vs-radians guards) close that gap cleanly. Good catch keeping the baseline consistent with `compute_angle_violation()`.

**Still outstanding from before:**
- **Rebase onto latest `main`** — GitHub still reports the branch as `BEHIND`; CONTRIBUTING asks for a rebase before merge.
- **CI on the new commit is still running** — `pytests`, `pre-commit-run`, and `security` are `pending` right now, so I can't confirm them green yet; worth a glance once they finish. (`pip-audit` remains the known repo-wide infra blocker — PYSEC-2026-3624 / lightning — not anything in your PR.)
- **Version bump** stays a maintainer call for @romeokienzler / @albanpuech.

Looking solid — a maintainer will follow up once CI settles.

 — 🤖 _automated pre-review; a maintainer will follow up_

**Comment by naomi-simumba** [comment]
🔴 1. Same unit bug in the AC/DC baseline

Fixed in `opf_ac_dc_baseline.py`. `_compute_branch_violations()` now converts `br["ang_min"]` and `br["ang_max"]` from degrees to radians before the comparisons, for both AC and DC angle violations. The conversion follows the same contract as `compute_angle_violation()` — a comment in the code makes this explicit.

🟡 2. Duplicated logic

The fix above consolidates the conversion into two shared variables (`ang_min_rad`, `ang_max_rad`) used by both the AC and DC blocks — so the degree→radian conversion is now done exactly once, and the comment ties it back to `compute_angle_violation()`. 
Additionally, two regression tests for` _compute_branch_violations` have also been added to `tests/test_violations.py` — `test_baseline_angle_violation_zero_within_limits` and `test_baseline_angle_violation_regression_degrees_vs_radians` — mirroring the existing tests for `compute_angle_violation` so both code paths are independently guarded against the same unit bug.

🟡 Denominator inconsistency

Added a clarifying note to the `compute_angle_violation()` docstring explicitly calling out that the mean is over all directed edges (2 × num_branches), which differs from the per-direction thermal violation denominators. Happy to align the denominators if that's the preferred direction — let me know and I'll update in this PR.

File: gridfm_graphkit/tasks/utils.py (hunk 1 above)
Comment (naomi-simumba) [reply gridfm_graphkit/tasks/utils.py:58]: Added a clarifying note to the compute_angle_violation() docstring explicitly calling out that the mean is over all directed edges (2 × num_branches), which differs from the per-direction thermal violation denominators. Happy to align the denominators if that's the preferred direction — let me know and I'll update in this PR.

**Comment by romeokienzler** [comment]
@naomi-simumba thanks — the new commit (`a94814c`) and your write-up close out the earlier points cleanly:

- **AC/DC baseline unit bug** — resolved. `_compute_branch_violations()` now converts `ang_min`/`ang_max` to radians once into `ang_min_rad`/`ang_max_rad`, used by both the AC and DC blocks, with a comment tying it back to `compute_angle_violation()`. ✔
- **Duplicated logic** — the shared `ang_min_rad`/`ang_max_rad` variables mean the conversion happens exactly once per path. ✔
- **Regression coverage** — the two new baseline tests (`test_baseline_angle_violation_zero_within_limits`, `test_baseline_angle_violation_regression_degrees_vs_radians`) mirror the `compute_angle_violation` tests, so both code paths are independently guarded against the unit bug. ✔

On your open **denominator question**: I'd lean toward *not* changing it in this PR. The angle metric is a documented, self-consistent choice (mean over all directed edges = 2 × num_branches), and your docstring now spells out exactly how it differs from the per-direction thermal metrics. Aligning the denominators is a metric-semantics change beyond this unit fix, so it's cleaner as its own follow-up if a maintainer wants it — happy to defer to @romeokienzler / @albanpuech on that.

**Still outstanding:**
- **Rebase onto latest `main`** — GitHub still reports the branch as `BEHIND`; CONTRIBUTING asks for a rebase before merge.
- **`pytests` is still `pending`** — worth a glance once it settles. Everything else is now green, including `pip-audit`, `pre-commit-run`, and `security` (the earlier `pip-audit` infra blocker has cleared).
- **Version bump** stays a maintainer call.

Nicely done — a maintainer will follow up once CI finishes.

 — 🤖 _automated pre-review; a maintainer will follow up_

**Comment by romeokienzler** [comment]
@naomi-simumba thanks — you've now rebased onto latest `main` (all commits replayed onto the current tip, and GitHub no longer reports the branch as `BEHIND`), which clears the last standing item from my earlier reviews. ✔

CI is re-running on the rebased commits — `pytests`, `pre-commit-run`, `security`, and `pip-audit` are `pending` right now, so it's just worth a glance once they settle green. The static scans (CodeQL, Trivy, Bandit, detect-secrets, DCO) are already passing.

Nothing else outstanding on my end. The only open thread is the **version bump**, which stays a maintainer call for @romeokienzler / @albanpuech. Looks ready for a maintainer's look once CI finishes.

 — 🤖 _automated pre-review; a maintainer will follow up_

**Comment by romeokienzler** [comment]
@naomi-simumba quick update after the fresh rebase onto today's \`main\`: the branch is no longer reported as \`BEHIND\`, and CI has now settled almost entirely green — \`pip-audit\`, \`security\`, \`pre-commit-run\`, and the static scans (CodeQL, Trivy, Bandit, detect-secrets, DCO) all pass. The only check still \`pending\` is \`pytests\`; just worth a glance once it finishes.

No new changes in the diff and nothing outstanding on my end. The single open thread remains the **version bump**, which stays a maintainer call for @romeokienzler / @albanpuech. Looks ready for a maintainer's look once \`pytests\` goes green.

 — 🤖 _automated pre-review; a maintainer will follow up_

**Review by romeokienzler — APPROVED** [review]
